### DISCUSSION TITLE
API input not being sanitized

### DISCUSSION THREAD

**Discussion by claytongulick** [directus/directus#15472] [Archive]
### Describe the Bug

I'm able to add malicious <script> tags to a textarea input, and it's not sanitized - it's happily stored in the database and retrieved as-is on query.

I would expect this input to be sanitized for all string types - am I missing a feature or setting somewhere?

While rendering, dompurify or similar can be used, shouldn't there be a sanitize option for these fields on the API side?

### To Reproduce

POST to any api endpoint and include a <script> tag or other in the text.

The malicious tag isn't sanitized when being stored.

### What version of Directus are you using?

9.14.5

### What version of Node.js are you using?

16.14.2

### What database are you using?

Postgres 14

### What browser are you using?

Chrome

### How are you deploying Directus?

local

**Comment by licitdev** [comment]
Input can be sanitised using the `filter event` hook, through flows or extensions.
Flows: https://docs.directus.io/configuration/flows/triggers.html#event-hook
Extensions: https://docs.directus.io/extensions/hooks.html#filter
Filter events: https://docs.directus.io/extensions/hooks.html#filter-events

Sanitisation is not enforced as some users may require to store and retrieve values containing `<script>` or values that are sensitive in certain context.

**Reply by claytongulick** [reply]
I think my concern with this is that sanitizing input in API endpoints is normally expected behavior - the golden rule of *never trust user input*. I'm not away of any endpoint I've ever written that didn't sanitize user input, it's normally a reportable critical security vulnerability in most systems I've seen.

It feels very strange to have to add an extension to sanitize input *for every single string field in the application*. If that's the case, why use Directus API functionality at all? If I have to write an extension for every table, how is that different from just writing my own express routes? I suppose there may be a way to do a generic one for all create/update events - but again, that feels like a very strange thing to have to do for a secure API.

I understand that we need to sanitize on rendering, but it seems like a pretty serious security vulnerability to me to have malicious user input directly stored in the database without any sanitation, I was pretty surprised when I was going through the source and couldn't find anything.

IMHO, sanitizing and scrubbing input should be the default behavior. For exceptions, it seems like having a field option similar to the slugify option for strings, like "allow raw values (warning - this could allow potentially malicious user input to be stored)".

OWASP recommendations: 
- https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html
- https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html

It looks like this could be addressed with just a few lines with: https://express-validator.github.io/docs/check-api.html#checkfields-message

**Reply by rijkvanzanten** [reply]
That really depends on the API that's being used, and what data is being ingested. I don't agree with that it's "normally expected behavior" that a data management layer API will automatically transform ingested data by default. As a matter of fact, automatically assuming certain data uses for the ingested data and applying (some) sanitization / transformation based on that would make the APIs IO very unpredictable. 

Directus' job is to allow you to store/retrieve/manage/visualize the data, but it doesn't care what that data is, means, or where it came from. The end user of Directus is responsible for where they gather the data from, and what they do with the data that's stored and retrieved from Directus.

Directus will _escape_ any input to ensure that you can't do any SQL injection and similar attacks, and it _will_ sanitize when previewing stored values in-app to prevent XSS and similar attacks, but it won't auto-transform ingested data by default. The end user of Directus is responsible for where they gather the data from, and what they do with the data that's stored and retrieved from Directus.

**Reply by rijkvanzanten** [reply]
To add on to that, it would also be impossible for us to make a single sanitization layer that would "just work". We can't just assume that all ingested data is stored in a single format (like HTML) that we can sanitize using some best practices, so the next best thing would be to run multiple sanitization layers for multiple expected types of data, that may or may not conflict and may or may not fit the data you're actually ingesting. Similarly, we don't know where the data is going to be used when it's retrieved from Directus, so now we're dealing with an infinite amount of cleanup that may or may not be required to pass "sanitization"[^1].

This would in turn cause a false sense of security, as the ingested data would only be sanitized if it fits a certain expected input from Directus. This to me would open up the door for way more unexpected security breaches, as you would now falsely assume that Directus will handle the sanitization of stored data.

[^1]: To give a tad exaggerated, but real example: Lets say you're ingesting JSON values into the API from an untrusted source. Your goal is to use this JSON data with the Stripe API. Is it now Directus' job to sanitize this input JSON to make sure it will fit Stripe's API input so that doesn't crash?

**Comment by claytongulick** [comment]
Hey @rijkvanzanten just as an FYI, I've been recommending Directus to a friend who's a CTO of a major DoD contractor ($500m+ contracts).

Other members of his team had also brought Directus up to him as well.

He had begun to explore it, but when he became aware of this issue he told me it was a deal killer.

I got a bit of egg on my face over it, because I had been talking up the security model.

**Reply by benhaynes** [reply]
Haha, breaking out the big guns of motivation! 😄 I won't comment on our docket (that's up to Rijk), but will mention that any 80/20 feature or fix can be sponsored to expedite. So things like this can be resolved expeditiously if they are a bottleneck to adoption. As with all other tickets, we'll get to them in time... but without being sponsored, we're doing what we can with our limited resources.

**Comment by Pachat** [comment]
Are Ditrctus inputs in general uses parameterized queries to prevent SQL injection?

**Reply by rijkvanzanten** [reply]
yes

**Comment by Pachat** [comment]
Great!

Le mar. 12 mars 2024, 20:21, Rijk van Zanten ***@***.***> a
écrit :
